camel/storages/vectordb_storages/milvus.py

```diff
@@ -0,0 +1,357 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+import logging
+import re
+from datetime import datetime
+from typing import Any, Dict, List, Optional, Tuple
+
+from camel.storages.vectordb_storages import (
+    BaseVectorStorage,
+    VectorDBQuery,
+    VectorDBQueryResult,
+    VectorDBStatus,
+    VectorRecord,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class MilvusStorage(BaseVectorStorage):
+    r"""An implementation of the `BaseVectorStorage` for interacting with
+    Milvus, a cloud-native vector search engine.
+
+    The detailed information about Milvus is available at:
+    `Milvus <https://milvus.io/docs/overview.md/>`_
+
+    Args:
+        vector_dim (int): The dimenstion of storing vectors.
+        url_and_api_key (Tuple[str, str]): Tuple containing
+            the URL and API key for connecting to a remote Milvus instance.
+            (default: :obj:`None`)
+        collection_name (Optional[str], optional): Name for the collection in
+            the Milvus. If not provided, set it to the current time with iso
+            format. (default: :obj:`None`)
+        **kwargs (Any): Additional keyword arguments for initializing
+            `MilvusClient`.
+
+    Raises:
+        ImportError: If `pymilvus` package is not installed.
+    """
+
+    def __init__(
+        self,
+        vector_dim: int,
+        url_and_api_key: Tuple[str, str],
+        collection_name: Optional[str] = None,
+        **kwargs: Any,
+    ) -> None:
+        try:
+            from pymilvus import MilvusClient
+        except ImportError as exc:
+            raise ImportError(
+                "Please install `pymilvus` first. You can install it by "
+                "running `pip install pymilvus`.") from exc
+
+        self._client: MilvusClient
+        self._create_client(url_and_api_key, **kwargs)
+        self.vector_dim = vector_dim
+        self.collection_name = (collection_name
+                                or self._generate_collection_name())
+        self._check_and_create_collection()
+
+    def _create_client(
+        self,
+        url_and_api_key: Tuple[str, str],
+        **kwargs: Any,
+    ) -> None:
+        r"""Initializes the Milvus client with the provided connection details.
+
+        Args:
+            url_and_api_key (Tuple[str, str]): The URL and API key for the
```

Comment:
update docstr as below:
 ```
       url_and_api_key (Tuple[str, str]): Tuple containing
            the URL and API key for connecting to a remote Milvus instance.
            URL maps to Milvus uri concepts, and it could be endpoint:port.
            API key can be username:pwd or api token if using zilliz cloud
            (hosted Milvus)
```
Let me know if you have better wording

Reply:
Plz use the following:
`      url_and_api_key (Tuple[str, str]): Tuple containing
           the URL and API key for connecting to a remote Milvus instance.
           URL maps to Milvus uri concept, typically "endpoint:port".
           API key maps to Milvus token concept, for self-hosted it's
           "username:pwd", for Zilliz Cloud (fully-managed Milvus) it's API Key.`